File: ch09/9_2.py
import random as random


def partition(a, p, r):
    x = a[r]
    i = p - 1
    for j in range(p, r):
        if a[j] <= x:
            i += 1
            a[i], a[j] = a[j], a[i]
    a[i + 1], a[r] = a[r], a[i + 1]
    return i + 1


def random_partition(a, p, r):
    i = random.randint(p, r)
    a[i], a[r] = a[r], a[i]
    return partition(a, p, r)


def random_select(a, p, r, i):
    if p == r:
        return a[p]
    q = random_partition(a, p, r)
    k = q - p + 1
    if i == k:
        return a[q]
    elif k > i:
        return random_select(a, p, q - 1, i)
    else:
        return random_select(a, q + 1, r, i - k)


def main():
    a = [0] * 10
    for i in range(0, 10):
        a[i] = random.randint(0, 100)
    print a
    print random_select(a, 0, 9, 2)


main()

File: ch09/9_2.cpp
#include <iostream>
#include <time.h>
using namespace std;
const n = 8;
int PARTITION(int A[], int p, int r);

int RANDOM(int p, int r){
	int t = rand()%(r - p + 1) + p;
	return t;
}

int RANDOMIZED_PARTITION(int A[], int p, int r){
	int i = RANDOM(p, r);
	swap(A[r], A[i]);
	return PARTITION(A, p, r);
}

int PARTITION(int A[], int p, int r){
	int x = A[r];
	int i = p - 1;
	for (int j = p; j <= r-1; j++){
		if (A[j] <= x){
			i++;
			swap(A[i] ,A[j]);
		}
	}
	swap(A[i + 1] ,A[r]);
	return i + 1;
}

int RANDOMIZED_SELECT(int A[], int p, int r, int i){
	if (p == r){
		return A[p];
	}
	int q = RANDOMIZED_PARTITION(A, p, r);
	int k = q - p + 1;
	if (i == k){
		return A[q];
	}
	else if(i < k){
		return RANDOMIZED_SELECT(A, p, q - 1, i);
	}
	else 
		return RANDOMIZED_SELECT(A, q + 1, r, i - k);
}

void main(){
	//int A[n] = {43,11,29,82,0,89};
	int A[n] = {0};
	srand( (unsigned)time( NULL ) );
	for (int i = 0; i < n; i++){
		A[i] = rand()%100;
		cout << A[i] << " ";
	}
	cout << endl;
	cout << RANDOMIZED_SELECT(A, 0, 7, 2) << endl;
}
